crates/light-client-prover/src/circuit.rs

```diff
@@ -61,23 +66,231 @@ pub fn run_circuit<DaV: DaVerifier, G: ZkvmGuest>(
-    // https://github.com/chainwayxyz/citrea/issues/1401
-    for journal in batch_proof_journals {
-        G::verify(&journal, &batch_proof_method_id.into()).unwrap();
-    }
+    // Deserialize batch proof journals
+    let deserialized_outputs: Vec<_> = input
+        .batch_proof_journals
+        .iter()
+        .map(|journal| {
+            G::verify_and_extract_output::<BatchProofCircuitOutput<DaV::Spec, [u8; 32]>>(
+                journal,
+                &input.batch_proof_method_id.into(),
+            )
+            .expect("Should have verified and extracted the batch proof")
+        })
+        .collect();
+
+    // Mapping from initial state root to final state root and last L2 height
+    let mut initial_to_final = std::collections::BTreeMap::<[u8; 32], ([u8; 32], u64)>::new();
+
+    let mut unchained_outputs = vec![];
+
+    let mut last_state_root = [0u8; 32];
+    let mut last_l2_height = 0;
+
+    // If we have a previous light client proof, check they can be chained
+    // If not, skip for now
+    // TODO: Once we have a manually planted light client proof use that and assume prev light client proof always exists
+    // So there will be no need for all these if lets
+    if let Some(previous_output) = &deserialized_previous_light_client_proof_journal {
+        last_l2_height = previous_output.last_l2_height;
+        last_state_root = previous_output.state_root;
+        for unchained_info in previous_output.unchained_batch_proofs_info.iter() {
+            if unchained_info.last_l2_height <= previous_output.last_l2_height {
+                continue;
+            }
+            // Add them directly as they are the ones that could not be matched
+            initial_to_final.insert(
+                unchained_info.initial_state_root,
+                (
+                    unchained_info.final_state_root,
+                    unchained_info.last_l2_height,
+                ),
+            );
+        }
+
+        for output in deserialized_outputs.iter() {
+            // Do not add if last l2 height is smaller or equal to previous output
+            // This is to defend against replay attacks, for example if somehow there is the script of batch proof 1 we do not need to go through it again
+            if output.last_l2_height <= previous_output.last_l2_height {
+                continue;
+            }
+            recursive_match_state_roots(
+                &mut initial_to_final,
+                &BatchProofInfo::new(
+                    previous_output.state_root,
+                    output.final_state_root,
+                    output.last_l2_height,
+                ),
+            );
+        }
+
+        // Do recursive matching for previous state root
+        recursive_match_state_roots(
+            &mut initial_to_final,
+            &BatchProofInfo::new(
+                previous_output.state_root,
+                previous_output.state_root,
```

Comment:
initial sr should be our genesis state root actually.

we can make this an optional input to the circuit.

we supply it in the first ever LC proof ourselves. then we output it on every proof.